.jenkins/jenkins.sh

```diff
@@ -0,0 +1,98 @@
+#!/bin/bash -f
+
+# This is the master script used to trigger Jenkins actions.
+# The idea of this script is to keep the amount of code in the "Execute shell" field small
+# Exmple syntax:
+# .jenkins/jenkins.sh test
+# Other actions such as test/build/deploy can be defined.
+
+### Some environment variables available from Jenkins
+### Note: for a complete list see https://jenkins.ginko.ch/env-vars.html
+# slave              The name of the build slave (daint, kesch, ...).
+# BUILD_NUMBER       The current build number, such as "153".
+# BUILD_ID           The current build id, such as "2005-08-22_23-59-59" (YYYY-MM-DD_hh-mm-ss).
+# BUILD_DISPLAY_NAME The display name of the current build, something like "#153" by default.
+# NODE_NAME          Name of the slave if the build is on a slave, or "master" if run on master.
+# NODE_LABELS        Whitespace-separated list of labels that the node is assigned.
+# JENKINS_HOME       The absolute path of the data storage directory assigned on the master node.
+# JENKINS_URL        Full URL of Jenkins, like http://server:port/jenkins/
+# BUILD_URL          Full URL of this build, like http://server:port/jenkins/job/foo/15/
+# JOB_URL            Full URL of this job, like http://server:port/jenkins/job/foo/
+
+exitError()
+{
+    echo "ERROR $1: $3" 1>&2
+    echo "ERROR     LOCATION=$0" 1>&2
+    echo "ERROR     LINE=$2" 1>&2
+    exit $1
+}
+
+# echo basic setup
+echo "####### executing: $0 $* (PID=$$ HOST=$HOSTNAME TIME=`date '+%D %H:%M:%S'`)"
+
+# start timer
+T="$(date +%s)"
+
+# check sanity of environment
+test -n "$1" || exitError 1001 ${LINENO} "must pass an argument"
+test -n "${slave}" || exitError 1005 ${LINENO} "slave is not defined"
+shortslave=`echo ${slave} | sed 's/[0-9]*$//g'`
+
+# some global variables
+action="$1"
+optarg="$2"
+
+# check presence of env directory
+pushd `dirname $0` > /dev/null
+envloc=`/bin/pwd`
+popd > /dev/null
+
+# Download the env
+. ${envloc}/env.sh
+
+# setup module environment and default queue
+test -f ${envloc}/env/machineEnvironment.sh || exitError 1201 ${LINENO} "cannot find machineEnvironment.sh script"
+. ${envloc}/env/machineEnvironment.sh
+
+# check that host (define in machineEnvironment.sh) and slave are consistent
+echo ${host} | grep "${shortslave}" || exitError 1006 ${LINENO} "host does not contain slave"
+
+# get root directory of where jenkins.sh is sitting
+root=`dirname $0`
+
+# act according to action specified
+case "${action}" in
+
+test )
+
+    # check if test script exists
+    script="${root}/test.sh"
```

Comment:
Between this and the definition of `root`, this is hard-coded to only run `test.sh`. Should this be parametrized in some way? For example, adding the `build` and `deploy` clauses even if they do not exist as `sh` files yet? Reading the comment at the top of the file, I'd expect that I can simply add a build/deploy script and call this with build/deploy (or possibly for any command), but as written it only accepts `test`.

Or you could set `script="${root}/${action}.sh"` to allow arbitrary actions to be defined. This will helpfully say the script does not exist if an invalid action is used.

Even better would be to have an `actions` subdirectory, so `script="${root}/actions/${action}.sh"`. This would avoid someone doing something silly like `jenkins.sh jenkins` which infinitely recurses, and also helps make it clearer that `jenkins.sh` is the entrypoint for these actions when looking at the directory structure.

Reply:
I like your suggestion. Implemented.